adapters/lora/src/main/java/org/eclipse/hono/adapter/lora/LoraCommandSubscriptionsHandler.java

```diff
@@ -0,0 +1,116 @@
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.adapter.lora;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.eclipse.hono.adapter.lora.providers.LoraProvider;
+import org.eclipse.hono.client.command.CommandConsumer;
+import org.eclipse.hono.notification.NotificationEventBusSupport;
+import org.eclipse.hono.notification.deviceregistry.LifecycleChange;
+import org.eclipse.hono.notification.deviceregistry.TenantChangeNotification;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.TriTuple;
+
+import io.opentracing.Span;
+import io.opentracing.Tracer;
+import io.opentracing.noop.NoopTracerFactory;
+import io.opentracing.tag.Tags;
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Context;
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+
+/**
+ * Class that holds the Lora adapter command subscriptions, that are shared between multiple {@link LoraProtocolAdapter}
+ * verticles. It also takes care to close the subscriptions for deleted/disabled tenants.
+ */
+public class LoraCommandSubscriptionsHandler implements LoraSubscriptionHandler {
+
+    private final Vertx vertx;
+    private final Tracer tracer;
+    private final Map<SubscriptionKey, TriTuple<CommandConsumer, LoraProvider, Context>> commandSubscriptions;
+
+    /**
+     * Creates a new Lora command subscription handler.
+     *
+     * @param vertx The Vert.x instance to use.
+     */
+    public LoraCommandSubscriptionsHandler(final Vertx vertx) {
+        this(vertx, null);
+    }
+
+    /**
+     * Creates a new Lora command subscription handler.
+     *
+     * @param vertx The Vert.x instance to use.
+     * @param tracer The tracer instance.
+     */
+    public LoraCommandSubscriptionsHandler(final Vertx vertx, final Tracer tracer) {
+        this.commandSubscriptions = new ConcurrentHashMap<>();
+        this.vertx = vertx;
+        this.tracer = tracer == null ? NoopTracerFactory.create() : tracer;
+    }
+
+    @Override
+    public void start() {
+        NotificationEventBusSupport.registerConsumer(vertx, TenantChangeNotification.TYPE,
+                notification -> {
+                    if (LifecycleChange.DELETE.equals(notification.getChange())
+                            || (LifecycleChange.UPDATE.equals(notification.getChange())
+                                    && !notification.isTenantEnabled())) {
+                        handleTenantTimeout(notification.getTenantId());
+                    }
+                });
```

Comment:
Before this PR, the LoraProtocolAdapter used the tenant-timeout mechanism as a hook to remove obsolete `commandSubscriptions` map entries. At the time this was implemented, the device registry notifications didn't exist.

Switching to a mechanism based on the device registry notifications now, we have more possibilities.
The command subscriptions here are associated with a tenant and a gateway device.
When such a gateway device entry is deleted or disabled, a corresponding `commandSubscriptions` map entry is also obsolete.
Therefore, it makes sense to also handle `DeviceChangeNotifications` here.